<h1 style='font-size:40px'> Handling Traffic Ingress Controllers</h1>

<h2 style='font-size:30px'> 
    A Quick Note on Ingresses
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            O último componente que resta em nosso ambiente é um objeto Ingress. 
        </li>
        <li>
            Ele será o balanceador de nossa aplicação. No nosso caso, usaremos uma <a href='https://github.com/kubernetes/ingress-nginx'>versão específica</a> desse objeto, que funciona com o Nginx.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    <a href='https://spacelift.io/blog/kubernetes-ingress'>
        Creating the Ingress Configuration
    </a>
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Aula em que configuramos o Ingress de nossa aplicação.
        </li>
        <li>
            De forma geral, estabelecemos um conjunto de critérios que definem para qual serviço o objeto encaminhará a sua requisição.
        </li>
    </ul>
 </div>

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ingress-service
  annotations:
    nginx.ingress.kubernetes.io/use-regex: 'true'
    nginx.ingress.kubernetes.io/rewrite-target: /$1
spec:
  ingressClassName: nginx
  rules:
    - http:
        paths:
          - path: /?(.*) 
            pathType: ImplementationSpecific
            backend:
              service:
                name: client-cluster-ip-service
                port: 
                  number: 3000
          - path: /api/?(.*)
            pathType: ImplementationSpecific
            backend:
              service:
                name: server-cluster-ip-service
                port: 
                  number: 5000

```
Exemplo de arquivo de configuração de um Ingress. Observe que o destino da requisição é definido com base no caminho informado.

<h3 style='font-size:30px;font-style:italic'> Campos Especiais do Ingress</h3>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Podemos observar que o arquivo acima possui certos campos que não costumavam ser declarados para os demais objetos. Por isso, vamos explicar abaixo o papel de cada um deles.
        </li>
    </ul>
 </div>

<h4 style='font-size:30px;font-style:italic;text-decoration:underline'> annotations</h4>
<div> 
    <ul style='font-size:20px'>
        <li> 
            Esse campo especifica algumas configurações gerais do nosso objeto. No nosso caso, permitimos a aplicação de RegEx sobre os paths das requisições, e os redefinimos como o primeiro padrão encontrado, antes de as encaminharmos ao objeto de destino.
        </li>
     </ul>
</div>

<h4 style='font-size:30px;font-style:italic;text-decoration:underline'> ingressClassName</h4>
<div> 
    <ul style='font-size:20px'>
        <li> 
            Nome do sub-tipo de classe Ingress a ser utilizada.
        </li>
     </ul>
</div>

<h4 style='font-size:30px;font-style:italic;text-decoration:underline'> pathType</h4>
<div> 
    <ul style='font-size:20px'>
        <li> 
            Se refere a como o Ingress interpretará o path de uma requisição como válido. 
        </li>
        <li>
            Existem três categorias admitidas para o `pathType`:
            <ul> 
                <li> 
                    <i> Exact</i>: O path informado deve se igualar examatamente à regra especificada. 
                </li>
                <li> 
                    <i> Prefix</i>: Opção mais flexível. O Ingress direcionará a requisição, caso algum prefixo se enquadre com as regras. 
                </li>
                <li> 
                    <i> ImplementationSpecific</i>: A validez da requisição é determinada pela IngressClass.
                </li>
            </ul>
        </li>
        <li>
            Abaixo, segue uma tabela com alguns exemplos que irão facilitar a compreensão:
        </li>
     </ul>
</div>

<table>
    <thead>
        <tr>
            <th>
                pathType            
            </th>
            <th>
                Path Rule
            </th>
            <th>
                Request Path (s)
            </th>
            <th>
                Matches?
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                Exact
            </td>
            <td>
                /foo
            </td>
            <td>
                /foo/
            </td>
            <td>
                No
            </td>
        </tr>
        <tr>
            <td>
                Prefix
            </td>
            <td>
                /aaa/bbb
            </td>
            <td>
                /aaa/bbbwyz
            </td>
            <td>
                No
            </td>
        </tr>
        <tr>
            <td>
                Prefix
            </td>
            <td>
                /, /aaa
            </td>
            <td>
                /aaa/ccc
            </td>
            <td>
                Yes, on account of /aaa
            </td>
        </tr>
        <tr>
            <td>
                Prefix
            </td>
            <td>
                /, /aaa/, /aaa/bbb 
            </td>
            <td>
                /ccc
            </td>
            <td>
                Yes, on account of /
            </td>
        </tr>
    </tbody>
</table>

In [11]:
! git add .
! git commit -m 'Solucionar o problema de acesso à aplicação via browse'
! git push

[master bc87ea1] Aula 247
 2 files changed, 206 insertions(+), 16 deletions(-)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 24 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 945 bytes | 945.00 KiB/s, done.
Total 4 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/felipesveiga/docker.git
   3e655c5..bc87ea1  master -> master


<p style='color:red'> Vi Aula 247 e 248; Não consigo acessar a aplicação do projeto via browser!; Solucionar o problema de acesso à aplicação via browser</p>